extends CharacterBody3D


const SPEED = 5.0
const JUMP_VELOCITY = 5.5
const SENSITIVITY = 0.003

const PLAYER_FREQ = 2.0
const PLAYER_AMP = 0.08
var t_player = 0.0

const OFFSET = 2

@onready var head = $Head
@onready var camera = $Head/Camera3D

func _ready():
	Input.set_mouse_mode(Input.MOUSE_MODE_CAPTURED)

func _unhandled_input(event):
	if event is InputEventMouseMotion:
		head.rotate_y(-event.relative.x * SENSITIVITY)
		camera.rotate_x(-event.relative.y * SENSITIVITY)
		camera.rotation.x = clamp(camera.rotation.x, deg_to_rad(-40), deg_to_rad(60))
		
		

func _physics_process(delta: float) -> void:
	# Add the gravity.
	if not is_on_floor():
		velocity += get_gravity() * delta

	# Handle jump.
	if Input.is_action_just_pressed("jump") and is_on_floor():
		velocity.y = JUMP_VELOCITY
		
	if Input.is_action_just_pressed("ui_cancel"):
		get_tree().quit()

	# Get the input direction and handle the movement/deceleration.
	# As good practice, you should replace UI actions with custom gameplay actions.
	var input_dir := Input.get_vector("move_left", "move_right", "move_forward", "move_backward")
	var direction = (head.transform.basis * Vector3(input_dir.x, 0, input_dir.y)).normalized()
	if direction:
		velocity.x = direction.x * SPEED
		velocity.z = direction.z * SPEED
	else:
		velocity.x = move_toward(velocity.x, 0, SPEED)
		velocity.z = move_toward(velocity.z, 0, SPEED)

	# Head player
	t_player += delta * velocity.length() * float(is_on_floor())
	camera.transform.origin = _headplayer(t_player)

	move_and_slide()

func _headplayer(time) -> Vector3:
	var pos = Vector3.ZERO
	pos.y = sin(time * PLAYER_FREQ) * PLAYER_AMP + OFFSET
	pos.x = cos(time * PLAYER_FREQ / 2) * PLAYER_AMP
	return pos
